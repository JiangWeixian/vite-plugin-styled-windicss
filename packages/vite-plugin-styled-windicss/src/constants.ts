export const APPLY_REGEX = /(.*)@apply([^`\$;]*);\n/gm
