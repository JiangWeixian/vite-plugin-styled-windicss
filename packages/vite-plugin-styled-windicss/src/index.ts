import type { Plugin } from 'vite'
import _debug from 'debug'
import { walk } from 'estree-walker'
import MagicString from 'magic-string'
import type { WindiPluginUtils } from '@windicss/plugin-utils'

const NAME = 'vite-plugin-styled-windicss'

const debug = {
  styledComponents: _debug(`${NAME}:transform:styledComponents`),
}

function VitePluginWindicss(): Plugin {
  let utils: WindiPluginUtils

  const plugin: Plugin = {
    name: NAME,
    enforce: 'post',
    async configResolved(config) {
      const windicss = config.plugins.find((i) => i.name === 'vite-plugin-windicss')
      utils = windicss!.api
    },
    async transform(code, id) {
      if (!utils) {
        return null
      }
      await utils.ensureInit()
      if (!utils.isDetectTarget(id) || !code.includes('styled-components')) return
      debug.styledComponents(id)
      const parsed = this.parse(code, {})
      let ms: MagicString
      walk(parsed, {
        enter: (node: any) => {
          if (node.type === 'TemplateElement' && node.value.cooked.includes('@apply')) {
            const next = node.value.cooked.replace(
              /(.*)@apply([^`$]*)\n/gm,
              (_match: string, pre: string, applyCss: string) => {
                const parsed = utils.transformCSS(`&{@apply ${applyCss}}`, id)
                return `${pre} ${parsed}`
              },
            )

            ms = ms || new MagicString(code)
            ms.overwrite(node.start, node.end, next)
          }
        },
      })
      if (ms!) {
        return {
          code: ms.toString(),
          map: ms.generateMap({
            file: id,
            includeContent: true,
            hires: true,
          }),
        }
      }
      return null
    },
  }

  return plugin
}

export default VitePluginWindicss
