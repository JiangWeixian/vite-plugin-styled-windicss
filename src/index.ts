import { Table } from './md'

Table()

export const welcome = () => {
  console.log('hello world')
}
