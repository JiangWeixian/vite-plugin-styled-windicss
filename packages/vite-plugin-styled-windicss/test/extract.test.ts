import { APPLY_REGEX } from '../src/constants'
import { cases } from './cases'

import { CSSParser } from 'windicss/utils/parser'
import Windi from 'windicss'

const processor = new Windi()

describe('extract', () => {
  it('should work for nested css', () => {
    const next = cases.nested.replace(APPLY_REGEX, (_match, pre, applyCss) => {
      const parser = new CSSParser(`&{@apply ${applyCss}}`, processor)
      return `${pre} ${parser.parse().build()}`
    })
    expect(next).toMatchSnapshot('nested-css')
  })
})
