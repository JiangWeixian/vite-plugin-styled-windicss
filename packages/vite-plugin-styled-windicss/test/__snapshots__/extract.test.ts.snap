// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract should work for nested css: nested-css 1`] = `
"
    const Group = styled.div\`
    [data-role=\\"aside-inner-content\\"] {
       & {
  display: -webkit-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -ms-flex-direction: column;
  -webkit-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -ms-flex-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  height: 100%;
}
      .todolist {
         & {
  display: -webkit-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -ms-flex-direction: column;
  -webkit-flex-direction: column;
  flex-direction: column;
  -webkit-box-align: center;
  -ms-flex-align: center;
  -webkit-align-items: center;
  align-items: center;
  width: 100%;
}      }
    }
  \`
  "
`;
